design/design-doc-webhooks.md

```diff
@@ -0,0 +1,383 @@
+abstraction but in `crossplane-runtime`.
+
+## Proposal
+
+There are four main pieces of a webhook:
+* Actual Go implementation of the logic.
+* Exposing an HTTP server accepting requests from Kubernetes API Server and
+  responding properly.
+* Handing off certificates to both API Server and the webhook server since TLS
+  is required.
+* Making sure the server is registered to Kubernetes API Server.
+
+Because of the limitations of the `controller-runtime` high-level abstraction,
+we will go with the low level abstraction and build the necessary structs in
+`crossplane-runtime` that will be used in providers.
+
+### The Abstraction Layer
+
+In `crossplane-runtime`, we will accept a list of functions for each type of
+the webhook and make sure they are called in given order whenever a request
+hits. The main difference between `controller-runtime` is that these functions
+do not have to be members of API structs; they can belong to any object or just
+be anonymous functions. Though the recommendation would be to have a struct
+similar to implementations of `ExternalClient` where all functions are provided
+by a single struct.
+
+### Exposing Webhook Server
+
+When a controller-runtime manager starts, the webhook server is also started
+automatically. So, as long as necessary implementations are there and hooked up
+with the main manager object, we have the server up and running. But in order to
+expose it to cluster, we need to create a `Service` resource similar to
+`Deployment` of the controller.
+
+The package manager will create an opinionated `Service` resource for the
+providers it installs and this will require Crossplane to have necessary RBAC
+for managing `Service` objects, too. Though it could be limited to which
+namespace it's installed, again similar to `Deployment` RBAC.
+
+### Certificate Distribution
+
+Kubernetes API server enforces use of TLS for the communication between the API
+server and the webhook server. The webhook server needs to have the TLS Key &
+Certificate and the API server needs to have a certificate bundle that is signed
+by that key. Normally, the controllers in the wild have to include a mechanism
+to generate the certificate to use in their webhook servers and that usually
+renders a bad UX for the admins since it either requires manual creation of the
+certificates or having `cert-manager` installed.
+
+We will utilize the fact that we are orchestrating the installation of provider
+controllers by our package manager. Crossplane will accept the TLS `Secret` as
+input to its installation. Then in every provider installation, it will mount
+this `Secret` to the provider container to use as the certificate & key. Since
+it also installs the necessary YAMLs that register the webhooks to API Server,
+it will inject the CA Bundle before the creation of those resources, which are
+`MutatingWebhookConfiguration`, `ValidatingWebhookConfiguration` and the
+`spec.webhook` part of `CustomResourceDefinition`s. 
```

Comment:
This also feels like it could be part of PRI -- I do acknowledge that it does not exist yet though, so this feedback isn't super helpful. If we think that is the direction we want to go though, this proposal could be a motivating use case.